src/pipelines/rust_app.rs

```diff
@@ -241,6 +236,39 @@ impl RustApp {
             wasm_output: hashed_wasm_name,
         })
     }
+
+    async fn wasm_opt_build(&self, hashed_name: &str) -> Result<()> {
+        // Zero means no optimizations (valid arg for wasm-opt), so we can skip calling wasm-opt as
+        // it wouldn't have any effect.
+        if self.wasm_opt.as_deref() == Some("0") {
+            return Ok(());
+        }
+
+        self.progress.set_message("calling wasm-opt");
+
+        // Ensure our output dir is in place.
+        let mode_segment = if self.cfg.release { "release" } else { "debug" };
+        let output = self.manifest.metadata.target_directory.join("wasm-opt").join(mode_segment);
+        fs::create_dir_all(&output).await.context("error creating wasm-opt output dir")?;
+
+        // Build up args for calling wasm-opt.
+        let output = output.join(hashed_name);
+        let arg_output = format!("--output={}", output.display());
+        let arg_opt_level = format!("-O{}", self.wasm_opt.as_deref().unwrap_or_default());
+        let target_wasm = self.cfg.staging_dist.join(hashed_name).to_string_lossy().to_string();
+        let args = vec![&arg_output, &arg_opt_level, &target_wasm];
+
+        // Invoke wasm-opt.
+        run_command("wasm-opt", &args).await?;
+
+        // Copy the generated WASM file to the dist dir.
+        self.progress.set_message("copying generated artifacts");
+        fs::copy(output, self.cfg.staging_dist.join(hashed_name))
+            .await
+            .context("error copying wasm file to dist dir")?;
```

Comment:
I would say that we should go ahead and remove the old unoptimized WASM from the staging dir just after this call.

Reply:
Sure 👍, was just following the existing code on this. Let's remove the old binary.
We could tell `wasm-opt` to overwrite the existing file as alternative but may break the file if something goes wrong inside `wasm-opt` eventually.